webapp/src/ts/modals/training-cards/training-cards.component.ts

```diff
@@ -0,0 +1,227 @@
+import { Component, NgZone, OnDestroy, OnInit } from '@angular/core';
+import { BsModalRef } from 'ngx-bootstrap/modal';
+import { Store } from '@ngrx/store';
+import { combineLatest, Subscription } from 'rxjs';
+
+import { MmModalAbstract } from '@mm-modals/mm-modal/mm-modal';
+import { XmlFormsService } from '@mm-services/xml-forms.service';
+import { EnketoService } from '@mm-services/enketo.service';
+import { Selectors } from '@mm-selectors/index';
+import { GlobalActions } from '@mm-actions/global';
+import { GeolocationService } from '@mm-services/geolocation.service';
+import { TranslateService } from '@mm-services/translate.service';
+import { TelemetryService } from '@mm-services/telemetry.service';
+
+@Component({
+  selector: 'training-cards-modal',
+  templateUrl: './training-cards.component.html'
+})
+export class TrainingCardsComponent extends MmModalAbstract implements OnInit, OnDestroy {
+
+  constructor(
+    bsModalRef: BsModalRef,
+    private ngZone: NgZone,
+    private store: Store,
+    private xmlFormsService: XmlFormsService,
+    private enketoService: EnketoService,
+    private geolocationService: GeolocationService,
+    private translateService: TranslateService,
+    private telemetryService: TelemetryService,
+  ) {
+    super(bsModalRef);
+    this.globalActions = new GlobalActions(this.store);
+  }
+
+  static id = 'training-cards-modal';
+  private geoHandle:any;
+  private globalActions;
+  formNoTitle = false;
+  form;
+  trainingForm;
+  formWrapperId = 'training-cards-form';
+  modalTitleKey = 'training_cards.modal.title';
+  loadingContent;
+  contentError;
+  hideModalFooter;
+  errorTranslationKey;
+  enketoError;
+  enketoStatus;
+  enketoSaving;
+  telemetryData;
+  showConfirmExit;
+  subscription: Subscription = new Subscription();
+
+  ngOnInit() {
+    this.telemetryData = { preRender: Date.now() };
+    this.reset();
+    this.subscribeToStore();
+  }
+
+  ngOnDestroy() {
+    this.subscription.unsubscribe();
+    this.geoHandle?.cancel();
+    // old code checked whether the component is reused before unloading the form
+    // this is because AngularJS created the new "controller" before destroying the old one
+    // in Angular, unless specific RouteReuseStrategies are employed, components are always
+    // destroyed before new ones are created
+    // see https://github.com/angular/angular/blob/10.2.x/packages/router/src/operators/activate_routes.ts#L37
+    // for Angular behavior
+    // see https://github.com/medic/cht-core/issues/2198#issuecomment-210202785 for AngularJS behavior
+    this.enketoService.unload(this.form);
+    this.globalActions.clearEnketoStatus();
```

Comment:
I think this has an interaction with loading other Enketo forms, where the new form gets loaded before this one gets unloaded. 
What I'm doing is: 
1. Open a report, hit edit
2. Hit back in the browser so I'm back on the report detail page
3. Make a change to a training form to trigger the modal to appear
4. After modal has appeared, hit browser back, so I get to the edit report page again
5. If I proceed to edit the form, I never get a `Submit` button on the last page, only a `Next` button that does nothing.

Reply:
There's a similar issue, when in the URL: `add/pregnancy_facility_visit_reminder` and then quickly loads `add/pregnancy`, my solution mitigates this scenario as well.

I created a deferred object that will be resolved when the enketo service has finished unloading, the enketo service's render won't run until the deferred object has been resolved. 

In the weird case that this deferred object doesn't resolve, then I've added a timeout to run the unload and resolve the deferred so the enketo service can finish rendering the new form.

Reply:
Commit with solution: https://github.com/medic/cht-core/pull/8017/commits/336a27e16d7047727354b94ed0dddedc75c4fa72